examples/tests/sdk_tests/resource/call.test.w

```diff
@@ -0,0 +1,221 @@
+bring expect;
+bring sim;
+
+enum MyEnum {
+  A,
+  B,
+  C
+}
+
+struct MyStruct {
+  field1: str;
+  field2: num;
+}
+
+inflight class MyResourceBackend impl sim.IResource {
+  var field1: str;
+  var field2: (str): str;
+  var ctx: sim.IResourceContext?;
+  new() {
+    this.field1 = "hello";
+    this.field2 = (name: str) => { return "Hello, " + name; };
+    this.ctx = nil;
+  }
+
+  pub onStart(ctx: sim.IResourceContext) {
+    this.ctx = ctx;
+  }
+  pub onStop() {}
+
+  pub throwsError() {
+    throw "Look ma, an error!";
+  }
+
+  pub printLogs() {
+    this.ctx!.log("an info log");
+    this.ctx!.log("another info log", sim.LogLevel.INFO);
+    this.ctx!.log("a debug log", sim.LogLevel.DEBUG);
+    this.ctx!.log("a verbose log", sim.LogLevel.VERBOSE);
+    this.ctx!.log("a warn log", sim.LogLevel.WARN);
+    this.ctx!.log("an error log", sim.LogLevel.ERROR);
+  }
+
+  pub methodNum(arg: num): num { return arg * 2; }
+  pub methodStr(arg: str): str { return arg + arg; }
+  pub methodBool(arg: bool): bool { return !arg; }
+  pub methodOptNum1(arg: num?): num? { return arg; }
+  pub methodOptNum2(arg: num?): num? { return nil; }
+  pub methodJson(arg: Json): Json { return { value: arg }; }
+  pub methodEnum(arg: MyEnum): MyEnum { assert(arg == MyEnum.A); return MyEnum.B; }
+  pub methodArray(arg: Array<num>): Array<num> { return arg.slice(1, arg.length - 1); }
+  pub methodMap(arg: Map<num>): Map<num> {
+    let var res = MutMap<num>{};
+    for key in arg.keys() {
+      res[key] = arg[key] * 2;
+    }
+    return res.copy();
+  }
+  pub methodStruct(arg: MyStruct): MyStruct {
+    return MyStruct{ field1: arg.field1 + arg.field1, field2: arg.field2 * 2 };
+  }
+  pub methodWithVariadics(arg1: str, ...arg2: Array<num>): num {
+    let var res = num.fromStr(arg1);
+    for arg in arg2 {
+      res += arg;
+    }
+    return res;
+  }
+  pub methodWithComplexTypes1(arg1: Array<MyStruct>): Map<num> {
+    return { "field1Length": arg1[0].field1.length };
+  }
+}
+
+class MyResource {
+  backend: sim.Resource;
+  new() {
+    this.backend = new sim.Resource(inflight () => {
+      return new MyResourceBackend();
+    });
```

Comment:
Is it possible to pass `ctx` to the factory?

```suggestion
    this.backend = new sim.Resource(inflight (ctx) => {
      return new MyResourceBackend(ctx);
    });
```

Reply:
Ah that's a good suggestion, let me try that -- should remove a bunch of ugliness if it works

Reply:
I guess if we take this idea to its logical conclusion than we can just rely on the constructor for starting resources and remove the `onStart` method, yeah?

Reply:
Done